File: analytic-collector/services/alikafka/kafka_producer.go
package kafkaproducer

import (
	"context"
	"errors"
	"fmt"
	"time"

	"github.com/confluentinc/confluent-kafka-go/kafka"
	"github.com/inconshreveable/log15"
)

// KafkaProducer holds Kafka producer configuration and client
type KafkaProducer struct {
	producer  *kafka.Producer
	topicName string
}

// KafkaConfig holds the necessary configurations for Kafka
type KafkaConfig struct {
	BootstrapServers string
	SASLUsername     string
	SASLPassword     string
	CALocation       string
	TopicName        string
}

// NewKafkaProducer initializes a new Kafka producer for Alibaba Cloud Kafka
func NewKafkaProducer(ctx context.Context, config KafkaConfig) (*KafkaProducer, error) {
	if config.BootstrapServers == "" || config.SASLUsername == "" || config.SASLPassword == "" || config.TopicName == "" {
		log15.Error("Invalid Kafka configuration provided")
		return nil, errors.New("missing required Kafka configuration")
	}

	producer, err := kafka.NewProducer(&kafka.ConfigMap{
		"bootstrap.servers":                     config.BootstrapServers,
		"sasl.mechanisms":                       "PLAIN",
		"security.protocol":                     "SASL_SSL",
		"ssl.ca.location":                       config.CALocation,
		"sasl.username":                         config.SASLUsername,
		"sasl.password":                         config.SASLPassword,
		"ssl.endpoint.identification.algorithm": "none", // Disable hostname verification for Alibaba Cloud Kafka
	})
	if err != nil {
		log15.Error("Failed to initialize Kafka producer: " + err.Error())
		return nil, err
	}

	kp := &KafkaProducer{
		producer:  producer,
		topicName: config.TopicName,
	}

	return kp, nil
}

// Close closes the Kafka producer connection
func (kp *KafkaProducer) Close() error {
	log15.Info("Closing Kafka producer...")
	kp.producer.Close()
	return nil
}

// Publish sends a message to the Kafka topic
func (kp *KafkaProducer) Publish(messageJSON []byte) error {
	deliveryChan := make(chan kafka.Event, 1)
	defer close(deliveryChan)

	err := kp.producer.Produce(&kafka.Message{
		TopicPartition: kafka.TopicPartition{Topic: &kp.topicName, Partition: kafka.PartitionAny},
		Value:          messageJSON,
	}, deliveryChan)

	if err != nil {
		log15.Error("Failed to send message: " + err.Error())
		return err
	}

	e := <-deliveryChan
	msg := e.(*kafka.Message)
	if msg.TopicPartition.Error != nil {
		log15.Error("Delivery failed: " + msg.TopicPartition.Error.Error())
		return msg.TopicPartition.Error
	}

	log15.Info("Message delivered to topic: " + *msg.TopicPartition.Topic)
	return nil
}

// FetchDataFromAPI simulates fetching data from an API (similar to the Python version)
func FetchDataFromAPI(apiURL string) ([]byte, error) {
	// Placeholder: Use actual HTTP requests to fetch from an API endpoint.
	// Example: Use "net/http" to send a GET request to apiURL and return response body.
	log15.Info(fmt.Sprintf("Fetching data from API: %s", apiURL))

	// For demonstration, return dummy JSON data
	return []byte(`{"example": "data from API"}`), nil
}

// Example of usage (optional: simulate main)
func Example() {
	ctx := context.Background()
	config := KafkaConfig{
		BootstrapServers: "your-kafka-bootstrap-servers",
		SASLUsername:     "your-kafka-username",
		SASLPassword:     "your-kafka-password",
		CALocation:       "your-ca-location",
		TopicName:        "your-topic-name",
	}

	// Initialize Kafka producer
	kafkaProducer, err := NewKafkaProducer(ctx, config)
	if err != nil {
		log15.Crit("Failed to initialize Kafka producer: ", err)
		return
	}
	defer kafkaProducer.Close()

	// Simulate API fetching and publishing loop
	apiURL := "http://127.0.0.1:8000/stream-data"
	for {
		data, err := FetchDataFromAPI(apiURL)
		if err != nil {
			log15.Error("Failed to fetch data from API: " + err.Error())
			time.Sleep(5 * time.Second)
			continue
		}

		err = kafkaProducer.Publish(data)
		if err != nil {
			log15.Error("Failed to publish data to Kafka: " + err.Error())
		}

		time.Sleep(5 * time.Second) // Simulate periodic publishing
	}
}

File: kafka-python-adv/producer/kafka_producer.py
from confluent_kafka import Producer, KafkaException
import json
import logging
import time
import setting

from fastapi import FastAPI, HTTPException

# Configure logging
logging.basicConfig(
    level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
)

# Kafka producer configuration
kafka_conf = setting.kafka_setting
oss_conf = setting.oss_setting
producer = Producer(
    {
        "bootstrap.servers": kafka_conf["bootstrap_servers"],
        "ssl.endpoint.identification.algorithm": "none",
        "sasl.mechanisms": "PLAIN",
        "ssl.ca.location": kafka_conf["ca_location"],
        "security.protocol": "SASL_SSL",
        "sasl.username": kafka_conf["sasl_plain_username"],
        "sasl.password": kafka_conf["sasl_plain_password"],
        "group.id": kafka_conf["group_name"],
        "auto.offset.reset": "latest",
        "fetch.message.max.bytes": "524288",  # 512 KB
    }
)

# Initiate 
app = FastAPI()



################################################################

# API endpoint
@app.post("/stream-data")
async def receive_stream_data(data):
    try:
        produce_data(data.dict())
    except Exception as e:
        raise HTTPException(
            status_code=500, detail=f"Failed to send data to Kafka: {e}"
        )
    return {"status": "success", "data_sent": data.dict()}

# Ensure all messages are flushed from Kafka producer on shutdown
@app.on_event("shutdown")
def shutdown_event():
    producer.flush()

################################################################

# Kafka producer
def delivery_report(err, msg):
    if err is not None:
        logging.error(f"Message delivery failed: {err}")
    else:
        logging.info(f"Message delivered to {msg.topic()} [{msg.partition()}]")


def produce_data(data:dict):
    try:
        timestamp = int(time.time())  # Add timestamp to data
        data["timestamp"] = timestamp
        
        message = json.dumps(data).encode("utf8")
        producer.produce(
            kafka_conf["topic"],
            message,
            callback=delivery_report,
        )
        producer.poll(0)
        logging.info(f"Data sent to topic {kafka_conf["topic"]}: {data}")
    except KafkaException as e:
        logging.error(f"Kafka error: {e}")
    except Exception as e:
        logging.error(f"Error producing data: {e}")


# if __name__ == "__main__":
#     produce_data(example_data)
